### PR TITLE
Support saving a multi-dataset ORSO file from the ISIS Reflectometry GUI

### LINKED ISSUE #36396: Include multiple datasets in ISIS Reflectometry ORSO ASCII file
The ISIS Reflectometry ORSO ASCII file implemented under #31073 should be extended to support including multiple datasets in a single file for the following circumstances:

- [ ] Include both stitched and unstitched (individual angle) data in a single output file.

- [ ] POLREF would like to have all the spin states saved into a single file, rather than having to output individual files for each spin state.

- [ ] Time sliced data is sometimes saved out but less frequently. It would be desirable to have the option to combine the sliced datasets into a single file, although this should be an option rather than being done automatically.

### PR DESCRIPTION
**Pull request by rbauststfc**
#### Summary of work

This PR updates the ISIS Reflectometry GUI to provide options for saving multiple datasets to a single ORSO file. The functionality actually performing the save was added to the `SaveISISReflectometryORSO` algorithm in PR #37157.

Fixes #36396. 

#### Further detail of work

- Checkbox `Save multiple datasets to a single file` has been added to allow users to choose whether to combine multiple datasets into a single file when auto-saving into the ORSO format. When an ORSO file is saved out that contains more than one dataset the suffix `_multi` is appended to the filename. As agreed with our scientists, the multi-dataset file will be limited to auto-saving in this first version - if workspaces are manually selected and saved via the Save ASCII tab they will go into separate ORSO files for now.

- A second new checkbox, `Additional columns (includes Q resolution)`, has been added to give the option to include additional data columns when saving non-stitched datasets into the ORSO format. Following discussion with our scientists, this PR also changes the default behaviour if for some reason the additional columns have been requested and cannot be calculated (an edge case that shouldn't be encountered via the GUI). In this case, a warning will now be logged and the file will be saved without them.

- I have included a couple of small fixes and tests for a couple of places where you could cause a crash when trying to perform actions in the save tab with names of workspaces that were no longer in the GUI.

- Relevant documentation has been updated.

- On the linked issue there is a final bullet point that has not been done in this PR (adding a right click option from the Runs tab batch table to manually save a group into a multi-dataset file). I will be opening a separate issue to look at this.

### To test:

1) Go to `Interfaces -> Reflectometry -> ISIS Reflectometry`.
1) Go to the Save ASCII tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO ASCII (*.ort)` from the drop-down.
1) In the Automatic Save section, tick the `Save as ASCII on completion` checkbox.
1) The `Include individual row outputs for groups` checkbox should be enabled. Select this checkbox.
1) The `Save multiple datasets to a single file` checkbox should be enabled and selected by default.
1) Go to the Runs tab, enter Investigation ID `1120015` and Cycle `11_3`. Click the Search button.
1) From the search results, click to highlight runs `11934`, `11935` and `11936` and click the Transfer button to transfer them into the batch table on the right hand side. Click the Process button to reduce the data.
1) If you navigate to the save path that you set in the Save ASCII tab, you should see a single `.ort` file with the suffix `_multi` in the filename. This file should contain multiple datasets. Assuming you have left the `Additional columns (includes Q resolution)` checkbox ticked, the non-stitched (individual angle) datasets should have 8 columns and the final stitched dataset should have 4 columns.
1) Go back to the Save ASCII tab and from the List of Workspaces section on the left, select workspaces `IvsQ_11934_11935+11936` and  `IvsQ_binned_11934`.
1) Click the big Save button on the right hand side to manually save to file. If you navigate to the save path that you set, you should see two separate `.ort` files with one dataset in each. This tests that the manual save is still outputting multiple datasets to separate files.
1) Still on the Save ASCII tab, in the Automatic Save section untick the `Include individual row outputs for groups` checkbox. This means that only the stitched dataset from a group multi-row reduction will be automatically saved out.
1) Go back to the Runs tab and re-process the group you just reduced.
1) If you navigate to the save path that you set in the Save ASCII tab you should see a single `.ort` file, but this time it should **not** have the suffix `_multi` in the filename. This file should contain only the stitched dataset and should have just 4 columns.
1) Test that the checkboxes on the Save ASCII tab are enabled and disabled correctly depending on the selected file format and auto-save options. The `Additional columns (includes Q resolution)` checkbox should only be enabled for the ORSO format. The `Save multiple datasets to a single file` checkbox should only be enabled if `Save as ASCII on completion` has been ticked and the ORSO file format has been selected. It should be disabled otherwise.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: docs/source/interfaces/reflectometry/ISIS Reflectometry.rst [hunk 1]
```diff
@@ -947,11 +948,26 @@ See :ref:`algm-SaveReflectometryAscii` for a description of the formats.
 |                               | to specify if you want a Header and/or Q Resolution  |
 |                               | column as well as specifying the delimiter.          |
 +-------------------------------+------------------------------------------------------+
+| ORSO Format Options           | When saving in the ORSO format you can specify if you|
+|                               | want a Q Resolution column. You can also use the     |
+|                               | ``Additional columns (includes Q resolution`` option |
```
Comment (cailafinn) [inline docs/source/interfaces/reflectometry/ISIS Reflectometry.rst:953]:
```suggestion
|                               | ``Additional columns (includes Q resolution)`` option|
```
Missing a bracket here

**Review by cailafinn — CHANGES_REQUESTED** [review]
Functionally, this works really well. Followed the instructions (with the Sample Data refl files as I didn't seem to have access to the Journal during testing). All working well, files were saved into the correct locations with the right prefixes and suffixes. Automatic saving worked well. Correct checkboxes disabled when settings adjusted. 

Built the docs, all looking good formatting and grammar-wise. 

Just have a few code notes. As discussed, as the functionality is good, happy for certain elements to be fixed in a maintenance issue to avoid delay. The build warning probably does need to be fixed though as Jenkins will start reporting those as failures soon.

File: qt/scientific_interfaces/ISISReflectometry/GUI/Save/AsciiSaver.cpp [hunk 2]
```diff
@@ -68,33 +73,54 @@ void AsciiSaver::runSaveAsciiAlgorithm(std::string const &savePath, std::string
...
 void AsciiSaver::save(std::string const &saveDirectory, std::vector<std::string> const &workspaceNames,
                       std::vector<std::string> const &logParameters, FileFormatOptions const &fileFormat) const {
+  if (!isValidSaveDirectory(saveDirectory)) {
+    throw InvalidSavePath(saveDirectory);
+  }
+
   // Setup the appropriate save algorithm
-  if (isValidSaveDirectory(saveDirectory)) {
+  if (shouldSaveToSingleFile(workspaceNames, fileFormat)) {
+    saveToSingleFile(workspaceNames, saveDirectory, logParameters, fileFormat);
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply qt/scientific_interfaces/ISISReflectometry/GUI/Save/AsciiSaver.cpp:118]: Thanks for this thought - on this one, would you be happy if I left it as it is? I'm a fan of returning early in situations where you've checked a condition and then there's no point continuing with the purpose of the method at all if that condition doesn't hold. When you're checking a condition to see if you should achieve the method's purpose in one way or another, though, then I personally prefer an if/else, as I find it clearer when reading to understand the intention of the code. I think my preference would be that there's flexibility on code style for this particular type of case, if that works for you.

File: qt/scientific_interfaces/ISISReflectometry/GUI/Save/AsciiSaver.cpp [hunk 3]
```diff
@@ -103,8 +129,30 @@ void AsciiSaver::save(std::string const &saveDirectory, std::vector<std::string>
...
+  if (!fileFormat.isORSOFormat()) {
+    return false;
+  }
+
+  if (workspaceNames.size() > 1) {
+    return true;
+  }
+
+  // If there is only one workspace name in the list then we may still have multiple datasets if it is a workspace group
+  const auto ws = workspace(workspaceNames.front());
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply qt/scientific_interfaces/ISISReflectometry/GUI/Save/AsciiSaver.cpp:150]: Agree, this is definitely maintenance as there's quite a bit of legacy code in here. I'll open an issue.

**Review by cailafinn — APPROVED** [review]
Functionally working well for the sample data. Produces multi-workspaces properly when auto-saving. 
Code all looking good to me now. East const/const west issue moved to maintenance. 
Docs looking good when built.

**Comment by sf1919** [comment]
@mantidproject/gatekeepers Checked the warnings and they are all unrelated to this PR so please go ahead and gatekeep.

**Comment by rbauststfc** [comment]
@mantidproject/gatekeepers it would be great if this one could be picked up by a gatekeeper, as it's blocking another piece of work that I'd like to put through review in time for code-freeze. Thanks!

**Pull request merged**